src/rocks_compaction_scheduler.cpp

```diff
@@ -28,27 +28,167 @@
+        wait();
+    }
+
+    namespace {
+        template <class T>
+        class unlock_guard {
+        public:
+          unlock_guard(T& lk) : lk_(lk) {
+            lk_.unlock();
+          }
+
+          ~unlock_guard() {
+            lk_.lock();
+          }
+
+          unlock_guard(const unlock_guard&) = delete;
+          unlock_guard& operator=(const unlock_guard&) = delete;
+
+        private:
+          T& lk_;
+        };
+    }
+
+    void CompactionBackgroundJob::run() {
+        Client::initThread(_name);
+        stdx::unique_lock<stdx::mutex> lk(_compactionMutex);
+        while (_compactionThreadRunning) {
+            // check if we have something to compact
+            if (_compactionQueue.empty())
+                _compactionWakeUp.wait(lk);
+            else {
+                // get item from queue
+                CompactOp op(std::move(_compactionQueue.front()));
+                _compactionQueue.pop_front();
+                // unlock mutex for the time of compaction
+                unlock_guard<decltype(lk)> rlk(lk);
+                // do compaction
+                op.doCompact(_db);
+            }
+        }
+        lk.unlock();
+        LOG(1) << "compaction thread terminating" << std::endl;
+    }
+
+    Status CompactionBackgroundJob::scheduleCompactOp(const std::string& begin, const std::string& end,
+                                                       bool rangeDropped, const std::function<void()>& cleanup) {
+        {
+            stdx::lock_guard<stdx::mutex> lk(_compactionMutex);
+            _compactionQueue.push_back({begin, end, rangeDropped, cleanup});
+        }
+        _compactionWakeUp.notify_one();
+        return Status::OK();
+    }
+
+    void CompactionBackgroundJob::CompactOp::doCompact(rocksdb::DB* db) const {
+        rocksdb::Slice _start_slice(_start_str);
+        rocksdb::Slice _end_slice(_end_str);
+
+        rocksdb::Slice* _start = !_start_str.empty() ? &_start_slice : nullptr;
+        rocksdb::Slice* _end = !_end_str.empty() ? &_end_slice : nullptr;
+
+        log() << "starting compaction of range: "
+              << (_start ? _start->ToString(true) : "<begin>") << " .. "
+              << (_end ? _end->ToString(true) : "<end>")
+              << " (_rangeDropped is " << _rangeDropped << ")";
+
+        if (_rangeDropped) {
+            auto s = rocksdb::DeleteFilesInRange(db, db->DefaultColumnFamily(), _start, _end);
+            if (!s.ok()) {
+                log() << "failed to delete files in range: " << s.ToString();
+            }
+        }
+
+        rocksdb::CompactRangeOptions compact_options;
+        compact_options.bottommost_level_compaction = rocksdb::BottommostLevelCompaction::kForce;
+        compact_options.exclusive_manual_compaction = false;
+        auto s = db->CompactRange(compact_options, _start, _end);
+        if (!s.ok()) {
+            log() << "failed to compact range: " << s.ToString();
+        }
+
+        if (_cleanup) {
```

Comment:
We shouldn't proceed with the cleanup handler if `db->CompactRange` returned error. So the check should contain `s.ok()`